3.optimizing_CP/CellProfiler_optuna_utils/optuna_profiling_utils.py

```diff
@@ -0,0 +1,1157 @@
+
+    Parameters
+    ----------
+    cell_count : list
+        A list of the cell counts for each well over time points.
+    target_cell_count : list
+        A list of the average cell counts for each well over time points.
+
+    Returns
+    -------
+    loss : float
+        The mean squared error between the cell counts and the target cell counts.
+    """
+    cell_count = torch.tensor(cell_count).float()
+    target_cell_count = torch.tensor(target_cell_count).float()
+
+    # calculate the mean squared error using torch
+    # mse = torch.nn.MSELoss()
+    mae = torch.nn.L1Loss()
+    loss = mae(cell_count, target_cell_count)
+    # loss = mse(cell_count, target_cell_count)
+    # convert the loss to a float
+    loss = loss.item()
+    return loss
+
+
+def harmonic_mean(array: np.array) -> int:
+    """This function calculates the harmonic mean of an array of values.
+
+    Returns
+    -------
+    Float
+        The harmonic mean of the array of values.
+    """
+    array_len = len(array)
+    sum_of_reciprocals = 0
+    for value in array:
+        if value == 0:
+            pass
+        else:
+            sum_of_reciprocals += 1 / value
+    if sum_of_reciprocals == 0:
+        harmonic_mean = 1000
+    else:
+        harmonic_mean = array_len / sum_of_reciprocals
+
+    return harmonic_mean
+
+
+def loss_function_from_CP_features(
+    profile_path: pathlib.Path,
+    feature_s_to_use: list,
+    loss_method: str = "harmonic_mean",
+) -> float:
+    """This is a loss function based directly on the features extracted from CellProfiler.
+
+    Parameters
+    ----------
+    profile_path : pathlib.Path
+        Path to the extracted pycytominer features from CellProfiler.
+    feature_s_to_use : list
+        feature list to use for the loss function.
+    loss_method : str, optional
+        Loss type to extract from features, by default "harmonic_mean"
+
+    Returns
+    -------
+    float
+        The calculated loss from the features extracted from CellProfiler.
+
+    Raises
+    ------
+    ValueError
+        If the loss method is not recognized in the conditional statement.
+    """
+
+    # read in the dataframe
+    df = pd.read_parquet(profile_path)
+    array_of_values = df[feature_s_to_use].to_numpy()
+    # flatten the array of values
+    array_of_values = array_of_values.flatten()
+
+    # calculate the harmonic mean of the array of values
+
+    def harmonic_mean(array: np.array) -> int:
+        """This function calculates the harmonic mean of an array of values.
+
+        Returns
+        -------
+        Float
+            The harmonic mean of the array of values.
+        """
+        array_len = len(array)
+        sum_of_reciprocals = 0
+        for value in array:
+            if value == 0:
+                pass
+            else:
+                sum_of_reciprocals += 1 / value
+        harmonic_mean = array_len / sum_of_reciprocals
+
+        return harmonic_mean
```

Comment:
It looks like this is the same function that was defined before the `loss_function_from_CP_features` starting at line 493. Recommend DNR here and calling the already define `harmonic_mean` function here instead of defining it again.

Reply:
addressed